### PR TITLE
Unicode (and [autoexec]) improvements

### PR DESCRIPTION
**Pull request by FeralChild64**
Since my previous [autoexec]-related PR was just merged, here is another one :) As the 1st commit is just a file formatting cleanup, and the 2nd one just adds Unicode-related definitions, I suggest to review this PR commit by commit.

-----

Unicode related improvement:
- the engine can now also convert strings from DOS code page to UTF8
- the engine can now handle non-normalized Unicode input

[autoexec] section improvements:
- `CONFIG -axadd <line>` command now converts the line from current DOS code page to UTF-8
- `CONFIG -axtype` now converts the [autoexec] section to currently set DOS code page

-----

Remarks:
- the easiest way to test non-normalized Unicode support is to alter `encode.sh` or `encode.bat`script (change `nfc` to `nfkd`), and execute it to generate incorrectly normalized translation files; such translations should now work
- the `Z:\AUTOEXEC.BAT` is not translated from UTF-8 to DOS code page yet (only `CONFIG -axtype` works with UTF-8); implementing this correctly needs some significant effort - my branch https://github.com/dosbox-staging/dosbox-staging/tree/fc/autoexec-1 mostly does it, but there are still some edge cases which are not resolved yet
- to test the `CONFIG -axtype`, try the following sequence of command:
```
KEYB pl 667
CONFIG -axadd ąćęłńóśż
CONFIG -axtype
KEYB pl 668
CONFIG -axtype
KEYB us 437
CONFIG -axtype
```
To type accented letters above - in keyboard layout `pl` press _Right ALT_+letter key combination. Explanation: both 667 and 668 are code pages for Polish language, but mutually incompatible. The 437 does not support Polish, most needed accented characters are absent.

### CODE REVIEW

**Comment by ThomasEricB** [comment]
This is pretty good @FeralChild64 !

File: include/string_utils.h [hunk 1]
```diff
@@ -293,23 +293,23 @@ std::string join_with_commas(const std::vector<std::string>& items,
...
+uint16_t get_utf8_code_page();
+
+// Convert the UTF-8 string to the format intended for display inside emulated
+// environment. Code page '0' means a pure 7-bit ASCII. Function without
+// code_page parameter uses current DOS code page.
+// Return value 'false' means there were problems with string decoding or
+// rendering, but the overall output should still be sane.
+bool utf8_to_dos(const std::string& str_in, std::string& str_out);
+bool utf8_to_dos(const std::string& str_in, std::string& str_out,
+                 const uint16_t code_page);
```
Comment (kcgen) [inline include/string_utils.h:308]: very helpful comment :+1:

File: contrib/resources/mapping/DECOMPOSITION.TXT [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+0x1ed8  0x004f 0x0323 0x0302  #LATIN CAPITAL LETTER O WITH CIRCUMFLEX AND DOT BELOW
+0x1ed9  0x006f 0x0323 0x0302  #LATIN SMALL LETTER O WITH CIRCUMFLEX AND DOT BELOW
+0x1eda  0x004f 0x031b 0x0301  #LATIN CAPITAL LETTER O WITH HORN AND ACUTE
+0x1edb  0x006f 0x031b 0x0301  #LATIN SMALL LETTER O WITH HORN AND ACUTE
+0x1edc  0x004f 0x031b 0x0300  #LATIN CAPITAL LETTER O WITH HORN AND GRAVE
+0x1edd  0x006f 0x031b 0x0300  #LATIN SMALL LETTER O WITH HORN AND GRAVE
+0x1ede  0x004f 0x031b 0x0309  #LATIN CAPITAL LETTER O WITH HORN AND HOOK ABOVE
+0x1edf  0x006f 0x031b 0x0309  #LATIN SMALL LETTER O WITH HORN AND HOOK ABOVE
+0x1ee0  0x004f 0x031b 0x0303  #LATIN CAPITAL LETTER O WITH HORN AND TILDE
+0x1ee1  0x006f 0x031b 0x0303  #LATIN SMALL LETTER O WITH HORN AND TILDE
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply contrib/resources/mapping/DECOMPOSITION.TXT:400]: Shh! The less users know, the better they sleep :D

File: src/misc/messages.cpp [hunk 3]
```diff
@@ -58,10 +58,11 @@ class Message {
 	                               std::map<uint16_t, std::string> &output_msg_by_codepage)
 	{
 		assert(msg.length());
-		const uint16_t cp = UTF8_GetCodePage();
+		const uint16_t cp = get_utf8_code_page();
 		if (output_msg_by_codepage[cp].empty()) {
-			if (!UTF8_RenderForDos(msg, output_msg_by_codepage[cp], cp))
+			if (!utf8_to_dos(msg, output_msg_by_codepage[cp], cp)) {
 				LOG_WARNING("LANG: Problem rendering string");
```
Comment (kcgen) [inline src/misc/messages.cpp:64]: maybe include the problematic strings when logging?

`LANG: Problem rendering string from '%s' to '%s'`

File: src/misc/messages.cpp [hunk 4]
```diff
@@ -88,12 +89,11 @@ class Message {
...
 
 		assert(rendered_msg.length());
-		const uint16_t cp = UTF8_GetCodePage();
+		const uint16_t cp = get_utf8_code_page();
 		if (rendered_msg_by_codepage[cp].empty()) {
-			if (!UTF8_RenderForDos(rendered_msg,
-			                       rendered_msg_by_codepage[cp],
-			                       cp))
+			if (!utf8_to_dos(rendered_msg, rendered_msg_by_codepage[cp], cp)) {
 				LOG_WARNING("LANG: Problem rendering string");
```
Comment (kcgen) [inline src/misc/messages.cpp:95]: (ditto)

`LANG: Problem rendering string from '%s' to '%s'`

File: src/misc/programs.cpp [hunk 5]
```diff
@@ -793,7 +793,9 @@ void CONFIG::Run(void)
 				return;
 			}
 			for (Bitu i = 0; i < pvars.size(); i++) {
-				sec->HandleInputline(pvars[i]);
+				std::string line_utf8 = {};
+				dos_to_utf8(pvars[i], line_utf8);
+				sec->HandleInputline(line_utf8);
 			}
```
Comment (kcgen) [inline src/misc/programs.cpp:799]: Maybe: `for (const auto& pvar : pvars)`

File: src/misc/unicode.cpp [hunk 6]
```diff
@@ -32,416 +32,618 @@
...
+	Grapheme() = default;
+	Grapheme(const uint16_t code_point);
 
-    bool IsEmpty() const;
-    bool IsValid() const;
-    bool HasMark() const;
-    uint16_t GetCodePoint() const;
+	bool IsEmpty() const;
+	bool IsValid() const;
+	bool HasMark() const;
```
Comment (kcgen) [inline src/misc/unicode.cpp:42]: Good catch fixing all this white-space formatting :+1:

File: src/misc/unicode.cpp [hunk 7]
```diff
@@ -32,416 +32,618 @@
...
-        { 0x20d0, 0x20ff }, // Combining Diacritical Marks for Symbols
-        { 0xfe20, 0xfe2f }, // Combining Half Marks
-    };
-
-    for (const auto &range : ranges)
-        if (code_point >= range.first && code_point <= range.second)
-            return true;
-
-    return false;
+	static const std::vector<std::pair<uint16_t, uint16_t>> ranges = {
```
Comment (kcgen) [inline src/misc/unicode.cpp:137]: Can go with a compile-time literal array:

`static constexpr std::pair<uint16_t, uint16_t> ranges[] = {`

 (the `for` loop can handle it too)

File: src/misc/unicode.cpp [hunk 8]
```diff
@@ -32,416 +32,618 @@
...
+	        {0x1ab0, 0x1aff}, // Combining Diacritical Marks Extended
+	        {0x1dc0, 0x1dff}, // Combining Diacritical Marks Supplement
+	        {0x20d0, 0x20ff}, // Combining Diacritical Marks for Symbols
+	        {0xfe20, 0xfe2f}, // Combining Half Marks
+	};
+
+	for (const auto& range : ranges) {
+		if (code_point >= range.first && code_point <= range.second) {
+			return true;
+		}
```
Comment (kcgen) [inline src/misc/unicode.cpp:149]: Just a comment; this is a nice use of `std::any_of`

*Is the code_point in any_of the ranges?*

```c++
auto in_range = [code_point](const auto& range) {
	return code_point >= range.first && code_point <= range.second;
};
return std::any_of(std::begin(ranges), std::end(ranges), in_range);
```

Even though I still find lambda's a bit hard on the eyes, the `any_of` algorithm is a more "code by intent" (and the compiler is then free to implement the search in parallel, etc).

File: src/misc/unicode.cpp [hunk 9]
```diff
@@ -32,416 +32,618 @@
...
+		return;
+	}
+
+	if (!is_combining_mark(code_point) || is_empty) {
+		// Not a combining mark or empty grapheme
+		Invalidate();
+		return;
+	}
+
+	if (std::find(marks.begin(), marks.end(), code_point) != marks.end()) {
```
Comment (kcgen) [inline src/misc/unicode.cpp:222]: Might as well pass `find` const iterators (`.cbegin()` and `.cend()`)

File: src/misc/unicode.cpp [hunk 10]
```diff
@@ -32,416 +32,618 @@
...
 
-    code_point = unknown_character;
-    mark_1st   = 0;
-    mark_2nd   = 0;
+	code_point = unknown_character;
+	marks.clear();
+	marks_sorted.clear();
 }
 
 void Grapheme::AddMark(const uint16_t code_point)
```
Comment (kcgen) [inline src/misc/unicode.cpp:209]: The incoming argument `code_point` is ambiguous with `Grapheme` own `code_point` member variable.

File: src/misc/unicode.cpp [hunk 11]
```diff
@@ -451,712 +653,878 @@ static bool wide_to_code_page(const std::vector<uint16_t> &str_in,
...
+
+		} else if (GCC_UNLIKELY(tokens[0] == "CODEPAGE")) {
+			auto check_no_code_page = [&](const uint16_t code_page) {
+				if ((new_config_mappings.find(code_page) !=
+				             new_config_mappings.end() &&
+				     new_config_mappings[code_page].valid) ||
+				    new_config_duplicates.find(code_page) !=
+				            new_config_duplicates.end()) {
+					error_code_page_defined(file_name_main,
+					                        line_num);
```
Comment (kcgen) [inline src/misc/unicode.cpp:1027]: Formatting (left alignment) in `if` criteria seem to be off

File: src/misc/unicode.cpp [hunk 12]
```diff
@@ -32,416 +32,618 @@
 
 CHECK_NARROWING();
```
Comment (kcgen) [inline src/misc/unicode.cpp:33]: Impressive that all of this is size-correct too; no small feat!

File: src/misc/unicode.cpp [hunk 13]
```diff
@@ -451,712 +653,878 @@ static bool wide_to_code_page(const std::vector<uint16_t> &str_in,
...
+			error_parsing(file_name, line_num);
+			return false;
+		}
+
+		if (GCC_UNLIKELY(tokens.size() == 1)) {
+			// Handle undefined character entry, ignore 7-bit ASCII
+			// codes
+			if (character_code >= decode_threshold_non_ascii) {
+				Grapheme grapheme;
+				add_if_not_mapped(new_mapping, character_code, grapheme);
```
Comment (kcgen) [inline src/misc/unicode.cpp:926]: I wasn't sure how this was being used because it goes out of scope.
Suggest: `const Grapheme grapheme; // placeholder`

File: src/misc/unicode.cpp [hunk 14]
```diff
@@ -451,712 +653,878 @@ static bool wide_to_code_page(const std::vector<uint16_t> &str_in,
...
-    code_page = static_cast<uint16_t>(tmp);
-    return true;
+	if (GCC_UNLIKELY(token.empty() || token.length() > 5)) {
+		return false;
+	}
+
+	for (size_t i = 0; i < token.size(); ++i) {
+		if (GCC_UNLIKELY(token[i] < '0' || token[i] > '9')) {
+			return false;
+		}
```
Comment (kcgen) [inline src/misc/unicode.cpp:782]: maybe direct:

```c++
	for (char c : token) {
		if (GCC_UNLIKELY(c < '0' || c > '9')) {
```

File: src/misc/unicode.cpp [hunk 15]
```diff
@@ -32,416 +32,618 @@
...
+	        0x221f, 0x2194, 0x25b2, 0x25bc, // 1c-1f
+	};
+
+	constexpr uint16_t codepoint_7f = 0x2302;
+
+	str_out.clear();
+	str_out.reserve(str_in.size());
+
+	for (const auto character : str_in) {
+		uint8_t byte = static_cast<uint8_t>(character);
```
Comment (kcgen) [inline src/misc/unicode.cpp:626]: `auto byte` (let RHS control type)

File: src/misc/unicode.cpp [hunk 16]
```diff
@@ -32,416 +32,618 @@
...
+	Grapheme(const uint16_t code_point);
 
-    bool IsEmpty() const;
-    bool IsValid() const;
-    bool HasMark() const;
-    uint16_t GetCodePoint() const;
+	bool IsEmpty() const;
+	bool IsValid() const;
+	bool HasMark() const;
+	uint16_t GetCodePoint() const;
```
Comment (kcgen) [inline src/misc/unicode.cpp:43]: Ensure return types are always checked (otherwise the caller isn't using the function right):

```c++
	[[nodiscard]] bool IsEmpty() const;
	[[nodiscard]] bool IsValid() const;
	[[nodiscard]] bool HasMark() const;
	[[nodiscard]] uint16_t GetCodePoint() const;
```

File: src/misc/unicode.cpp [hunk 17]
```diff
@@ -32,416 +32,618 @@
...
-           mark_2nd   == other.mark_2nd &&
-           is_empty   == other.is_empty &&
-           is_valid   == other.is_valid);
-
-    return false;
+	if (code_point < other.code_point) {
+		return true;
+	} else if (code_point > other.code_point) {
+		return false;
+	}
```
Comment (kcgen) [inline src/misc/unicode.cpp:267]: The `else if` ca become a stand-alone `if` check because the prior `if` returns.

File: src/misc/unicode.cpp [hunk 18]
```diff
@@ -32,416 +32,618 @@
...
+
+	if (marks_sorted.empty() && other.marks_sorted.empty()) {
+		return false;
+	}
+
+	if (marks_sorted.size() < other.marks_sorted.size()) {
+		return true;
+	} else if (marks_sorted.size() > other.marks_sorted.size()) {
+		return false;
+	}
```
Comment (kcgen) [inline src/misc/unicode.cpp:277]: The `else if` can become a stand-alone `if` check because the prior `if` returns.

File: src/misc/unicode.cpp [hunk 19]
```diff
@@ -32,416 +32,618 @@
...
+
+	for (size_t idx = 0; idx < marks_sorted.size(); ++idx) {
+		if (marks_sorted[idx] < other.marks_sorted[idx]) {
+			return true;
+		} else if (marks_sorted[idx] > other.marks_sorted[idx]) {
+			return false;
+		}
```
Comment (kcgen) [inline src/misc/unicode.cpp:284]: The `else if` can become a stand-alone `if` because the prior `if` returns.

File: src/misc/unicode.cpp [hunk 20]
```diff
@@ -451,712 +653,878 @@ static bool wide_to_code_page(const std::vector<uint16_t> &str_in,
...
 static uint16_t deduplicate_code_page(const uint16_t code_page)
 {
-    const auto it = config_duplicates.find(code_page);
+	const auto it = config_duplicates.find(code_page);
+
+	if (it == config_duplicates.end()) {
+		return code_page;
+	} else {
+		return it->second;
+	}
```
Comment (kcgen) [inline src/misc/unicode.cpp:1270]: The `else if` can become a stand-alone `if` because the prior `if` returns.

File: src/misc/unicode.cpp [hunk 21]
```diff
@@ -451,712 +653,878 @@ static bool wide_to_code_page(const std::vector<uint16_t> &str_in,
...
+
+	load_config_if_needed();
+	const uint16_t code_page = deduplicate_code_page(dos.loaded_codepage);
+
+	// For unsupported code pages revert to default one
+	if (prepare_code_page(code_page)) {
+		return code_page;
+	} else {
+		return get_default_code_page();
+	}
```
Comment (kcgen) [inline src/misc/unicode.cpp:1475]: The `else ... return` can be flatten away to just `return`.

**Comment by kcgen** [comment]
Despite the complexity necessary, the code is extremely natural to read with the very clear variable names, bool names, and extensive self-documenting lambas (along with all the error logging in the labdas to avoid reptition).  This quite a masterpeice of unicode handling @FeralChild64!

I'm not sure how to type the Polish characters (as I routinely don't use the `Alt+combos), however at least testing w/ the sanitizers and passing in batch files is OK:



If there are any copy-and-paste (or bat files) I can help test with, happy to run them through if you can attach any.

**Comment by FeralChild64** [comment]
@kcgen 

> I'm not sure how to type the Polish characters

Right ALT + letter (has to be the right one, left ALT won't work).

File: src/misc/unicode.cpp (hunk 16 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: Ok, changed.

File: src/misc/unicode.cpp (hunk 15 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: Changed to `const auto byte`.

File: src/misc/unicode.cpp (hunk 14 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: Ok, changed to range loop.

File: src/misc/unicode.cpp (hunk 9 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: Ok, changed.

File: src/misc/programs.cpp (hunk 5 above)
Comment (FeralChild64) [reply src/misc/programs.cpp]: Ok, changed to range loop.

File: src/misc/unicode.cpp (hunk 21 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: Done.

File: src/misc/unicode.cpp (hunk 20 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: Done.

File: src/misc/unicode.cpp (hunk 19 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: Done.

File: src/misc/unicode.cpp (hunk 18 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: Done.

File: src/misc/unicode.cpp (hunk 17 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: Done.

File: src/misc/unicode.cpp (hunk 10 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: Changed argument to `in_code_point` - I have seen such a convention in DOSBox Staging, and I think it is sane.

File: src/misc/messages.cpp (hunk 3 above)
Comment (FeralChild64) [reply src/misc/messages.cpp]: Changed, but I'm tracing only the source string. Destination string might contain something the host OS (if it is UTF-8 aware) does not know how to handle.

File: src/misc/messages.cpp (hunk 4 above)
Comment (FeralChild64) [reply src/misc/messages.cpp]: Changed, but I'm tracing only the source string. Destination string might contain something the host OS (if it is UTF-8 aware) does not know how to handle.

File: src/misc/unicode.cpp (hunk 11 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: - 8 character long tabulation
- 80 characters line limit
- policy of using very long, descriptive variable names
- C++17, where you often have to write very long statements to use modern library features
and the clang-format script produced something like this. Reformatted the code manually.

File: src/misc/unicode.cpp [hunk 22]
```diff
@@ -451,712 +653,878 @@ static bool wide_to_code_page(const std::vector<uint16_t> &str_in,
...
+			}
+
+			auto& new_mapping =
+			        new_config_mappings[curent_code_page].mapping;
+
+			if (tokens.size() == 1) {
+				// Handle undefined character entry
+				if (character_code >= decode_threshold_non_ascii) {
+					// ignore 7-bit ASCII codes
+					Grapheme grapheme;
```
[the remark this answers is not in the document]
Comment (FeralChild64) [reply src/misc/unicode.cpp]: Ok, changed.

File: src/misc/unicode.cpp (hunk 13 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp:929]: Ok, changed.

File: src/misc/unicode.cpp (hunk 7 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: OK, changed.

File: src/misc/unicode.cpp (hunk 8 above)
Comment (FeralChild64) [reply src/misc/unicode.cpp]: Ok, changed.

File: src/misc/unicode.cpp (hunk 10 above)
Comment (kcgen) [reply src/misc/unicode.cpp]: Thanks for noting this standard; makes sense and will try to follow as well!

File: src/misc/unicode.cpp (hunk 11 above)
Comment (kcgen) [reply src/misc/unicode.cpp]: (oh, and I see these were formatted inside the nested parentheses; might have already been OK in that regard! Sorry if this comment wasn't helpful!)

**Review by kcgen — APPROVED** [review]
Time to merge; all comments addressed and passing tests. 🚀
